examples/aco_tsp/aco_tsp/model.py

```diff
@@ -0,0 +1,249 @@
+            v_x, v_y = self.g.nodes[v]["pos"]
+            self.g[u][v]["distance"] = ((u_x - v_x) ** 2 + (u_y - v_y) ** 2) ** 0.5
+            self.g[u][v]["visibility"] = 1 / self.g[u][v]["distance"]
+            self.g[u][v]["pheromone"] = self.pheromone_init
+
+    @classmethod
+    def from_random(cls, num_cities: int, seed: int = 0) -> "TSPGraph":
+        g = nx.random_geometric_graph(num_cities, 2.0, seed=seed).to_directed()
+
+        return cls(g)
+
+    @classmethod
+    def from_tsp_file(cls, file_path: str) -> "TSPGraph":
+        with open(file_path) as f:
+            lines = f.readlines()
+            # Skip lines until reach the text "NODE_COORD_SECTION"
+            while lines.pop(0).strip() != "NODE_COORD_SECTION":
+                pass
+
+            g = nx.Graph()
+            for line in lines:
+                if line.strip() == "EOF":
+                    break
+                node_coordinate = NodeCoordinates.from_line(line)
+
+                g.add_node(
+                    node_coordinate.city, pos=(node_coordinate.x, node_coordinate.y)
+                )
+
+        # Add edges between all nodes to make a complete graph
+        for u in g.nodes():
+            for v in g.nodes():
+                if u == v:
+                    continue
+                g.add_edge(u, v)
+
+        return cls(g)
+
+
+class AntTSP(mesa.Agent):
+    """
+    An agent
+    """
+
+    def __init__(self, unique_id, model, alpha: float = 1.0, beta: float = 5.0):
+        """
+        Customize the agent
+        """
+        self.unique_id = unique_id
+        self.alpha = alpha
+        self.beta = beta
+        super().__init__(unique_id, model)
+        self._cities_visited = []
+        self._traveled_distance = 0
+        self.tsp_solution = []
+        self.tsp_distance = 0
+
+    def calculate_pheromone_delta(self, q: float = 100):
+        results = {}
+        for idx, start_city in enumerate(self.tsp_solution[:-1]):
+            end_city = self.tsp_solution[idx + 1]
+            results[(start_city, end_city)] = q / self.tsp_distance
+
+        return results
+
+    def decide_next_city(self):
+        # Random
+        # new_city = self.random.choice(list(self.model.all_cities - set(self.cities_visited)))
+        # Choose closest city not yet visited
+        g = self.model.grid.G
+        current_city = self.pos
+        neighbors = list(g.neighbors(current_city))
+        candidates = [n for n in neighbors if n not in self._cities_visited]
+        if len(candidates) == 0:
+            return current_city
+
+        # p_ij(t) = 1/Z*[(tau_ij)**alpha * (1/distance)**beta]
+        results = []
+        for city in candidates:
+            val = (
+                (g[current_city][city]["pheromone"]) ** self.alpha
+                * (g[current_city][city]["visibility"]) ** self.beta
+            )
+            results.append(val)
+
+        results = np.array(results)
+        Z = results.sum()
```

Comment:
It's a bit uncommon to use a capital letter for a variable (it's usually reserved for classes).